File: archive/typescript-version/src/core/todo.manager.ts
import { promises as fs } from 'fs';
import { join } from 'path';
import { TodoItem } from '../types';
import chalk from 'chalk';

export class TodoManager {
  private todoDir: string;

  constructor(outputDir = './reviews') {
    this.todoDir = join(outputDir, 'todo');
  }

  async getAllTodos(): Promise<TodoItem[]> {
    try {
      await fs.access(this.todoDir);
    } catch {
      await fs.mkdir(this.todoDir, { recursive: true });
      return [];
    }

    try {
      const files = await fs.readdir(this.todoDir);
      const todoFiles = files.filter(f => f.startsWith('todo-') && f.endsWith('.json'));
      
      const allTodos: TodoItem[] = [];
      
      for (const file of todoFiles) {
        const filepath = join(this.todoDir, file);
        const content = await fs.readFile(filepath, 'utf-8');
        const todos = JSON.parse(content) as TodoItem[];
        allTodos.push(...todos);
      }
      
      return allTodos.sort((a, b) => {
        if (a.completed && !b.completed) return 1;
        if (!a.completed && b.completed) return -1;
        
        const priorityOrder = { high: 0, medium: 1, low: 2 };
        const aPriority = priorityOrder[a.priority] ?? 1;
        const bPriority = priorityOrder[b.priority] ?? 1;
        
        if (aPriority !== bPriority) {
          return aPriority - bPriority;
        }
        
        return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
      });
    } catch (error) {
      console.warn('Error reading TODO items:', error);
      return [];
    }
  }

  async getActiveTodos(): Promise<TodoItem[]> {
    const allTodos = await this.getAllTodos();
    return allTodos.filter(todo => !todo.completed);
  }

  async getCompletedTodos(): Promise<TodoItem[]> {
    const allTodos = await this.getAllTodos();
    return allTodos.filter(todo => todo.completed);
  }

  async markTodoCompleted(todoId: string): Promise<void> {
    const todos = await this.getAllTodos();
    const todo = todos.find(t => t.id === todoId);
    
    if (!todo) {
      throw new Error(`TODO item with id '${todoId}' not found`);
    }
    
    if (todo.completed) {
      console.log(`TODO item '${todoId}' is already completed`);
      return;
    }
    
    todo.completed = true;
    await this.saveTodoUpdate(todo);
    
    console.log(chalk.green(`✓ TODO item '${todo.title}' marked as completed`));
  }

  async markTodoActive(todoId: string): Promise<void> {
    const todos = await this.getAllTodos();
    const todo = todos.find(t => t.id === todoId);
    
    if (!todo) {
      throw new Error(`TODO item with id '${todoId}' not found`);
    }
    
    if (!todo.completed) {
      console.log(`TODO item '${todoId}' is already active`);
      return;
    }
    
    todo.completed = false;
    await this.saveTodoUpdate(todo);
    
    console.log(chalk.blue(`↺ TODO item '${todo.title}' marked as active`));
  }

  async deleteTodo(todoId: string): Promise<void> {
    const todos = await this.getAllTodos();
    const todoIndex = todos.findIndex(t => t.id === todoId);
    
    if (todoIndex === -1) {
      throw new Error(`TODO item with id '${todoId}' not found`);
    }
    
    const todo = todos[todoIndex];
    todos.splice(todoIndex, 1);
    
    await this.saveAllTodos(todos);
    
    console.log(chalk.red(`🗑 TODO item '${todo.title}' deleted`));
  }

  async displayTodos(options: {
    showCompleted?: boolean;
    filter?: 'all' | 'active' | 'completed';
    priority?: 'high' | 'medium' | 'low';
  } = {}): Promise<void> {
    let todos = await this.getAllTodos();
    
    if (options.filter === 'active') {
      todos = todos.filter(t => !t.completed);
    } else if (options.filter === 'completed') {
      todos = todos.filter(t => t.completed);
    }
    
    if (options.priority) {
      todos = todos.filter(t => t.priority === options.priority);
    }

    if (todos.length === 0) {
      console.log(chalk.gray('No TODO items found'));
      return;
    }

    const activeTodos = todos.filter(t => !t.completed);
    const completedTodos = todos.filter(t => t.completed);

    if (activeTodos.length > 0) {
      console.log(chalk.bold('\n📝 Active TODOs:\n'));
      this.displayTodoList(activeTodos);
    }

    if (completedTodos.length > 0 && (options.showCompleted || options.filter === 'completed')) {
      console.log(chalk.bold('\n✅ Completed TODOs:\n'));
      this.displayTodoList(completedTodos, true);
    }

    this.displaySummary(activeTodos.length, completedTodos.length);
  }

  private displayTodoList(todos: TodoItem[], showCompleted = false): void {
    todos.forEach((todo, index) => {
      const priorityIcon = this.getPriorityIcon(todo.priority);
      const statusIcon = todo.completed ? '✅' : '⭕';
      const todoText = showCompleted ? chalk.strikethrough(todo.title) : todo.title;
      
      console.log(`${statusIcon} ${priorityIcon} [${chalk.cyan(todo.id)}] ${todoText}`);
      
      if (todo.description && todo.description !== todo.title) {
        console.log(chalk.gray(`   ${todo.description}`));
      }
      
      if (todo.files && todo.files.length > 0) {
        console.log(chalk.gray(`   📁 Files: ${todo.files.join(', ')}`));
      }
      
      const createdDate = new Date(todo.createdAt).toLocaleDateString();
      console.log(chalk.gray(`   📅 Created: ${createdDate} | Review: ${todo.reviewId}`));
      
      if (todo.dueDate) {
        const dueDate = new Date(todo.dueDate).toLocaleDateString();
        console.log(chalk.gray(`   ⏰ Due: ${dueDate}`));
      }
      
      console.log();
    });
  }

  private getPriorityIcon(priority: string): string {
    switch (priority) {
      case 'high': return chalk.red('🔴');
      case 'medium': return chalk.yellow('🟡');
      case 'low': return chalk.green('🟢');
      default: return '⚪';
    }
  }

  private displaySummary(active: number, completed: number): void {
    const total = active + completed;
    const completionRate = total > 0 ? Math.round((completed / total) * 100) : 0;
    
    console.log(chalk.bold('\n📊 Summary:'));
    console.log(`Total: ${chalk.blue(total.toString())} | Active: ${chalk.yellow(active.toString())} | Completed: ${chalk.green(completed.toString())} | Progress: ${chalk.cyan(completionRate + '%')}`);
  }

  private async saveTodoUpdate(todo: TodoItem): Promise<void> {
    const todos = await this.getAllTodos();
    const index = todos.findIndex(t => t.id === todo.id);
    
    if (index !== -1) {
      todos[index] = todo;
      await this.saveAllTodos(todos);
    }
  }

  private async saveAllTodos(todos: TodoItem[]): Promise<void> {
    const files = await fs.readdir(this.todoDir);
    const todoFiles = files.filter(f => f.startsWith('todo-') && f.endsWith('.json'));
    
    for (const file of todoFiles) {
      const filepath = join(this.todoDir, file);
      await fs.unlink(filepath);
    }

    const todosByDate: { [date: string]: TodoItem[] } = {};
    
    todos.forEach(todo => {
      const dateKey = todo.createdAt.split('T')[0];
      if (!todosByDate[dateKey]) {
        todosByDate[dateKey] = [];
      }
      todosByDate[dateKey].push(todo);
    });

    for (const [date, dateTodos] of Object.entries(todosByDate)) {
      const filename = `todo-${date}.json`;
      const filepath = join(this.todoDir, filename);
      await fs.writeFile(filepath, JSON.stringify(dateTodos, null, 2), 'utf-8');
    }
  }

  async exportTodos(format: 'json' | 'csv' | 'markdown' = 'json', outputPath?: string): Promise<void> {
    const todos = await this.getAllTodos();
    
    if (todos.length === 0) {
      console.log('No TODOs to export');
      return;
    }

    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const defaultPath = `todos-export-${timestamp}`;
    
    switch (format) {
      case 'json':
        const jsonPath = outputPath || `${defaultPath}.json`;
        await fs.writeFile(jsonPath, JSON.stringify(todos, null, 2), 'utf-8');
        console.log(chalk.green(`TODOs exported to: ${jsonPath}`));
        break;
        
      case 'csv':
        const csvPath = outputPath || `${defaultPath}.csv`;
        const csvContent = this.convertToCsv(todos);
        await fs.writeFile(csvPath, csvContent, 'utf-8');
        console.log(chalk.green(`TODOs exported to: ${csvPath}`));
        break;
        
      case 'markdown':
        const mdPath = outputPath || `${defaultPath}.md`;
        const mdContent = this.convertToMarkdown(todos);
        await fs.writeFile(mdPath, mdContent, 'utf-8');
        console.log(chalk.green(`TODOs exported to: ${mdPath}`));
        break;
    }
  }

  private convertToCsv(todos: TodoItem[]): string {
    const headers = ['ID', 'Title', 'Description', 'Priority', 'Status', 'Files', 'Created', 'Due Date', 'Review ID'];
    const rows = todos.map(todo => [
      todo.id,
      `"${todo.title.replace(/"/g, '""')}"`,
      `"${(todo.description || '').replace(/"/g, '""')}"`,
      todo.priority,
      todo.completed ? 'Completed' : 'Active',
      `"${(todo.files || []).join(', ')}"`,
      new Date(todo.createdAt).toLocaleDateString(),
      todo.dueDate ? new Date(todo.dueDate).toLocaleDateString() : '',
      todo.reviewId
    ]);
    
    return [headers.join(','), ...rows.map(row => row.join(','))].join('\n');
  }

  private convertToMarkdown(todos: TodoItem[]): string {
    const active = todos.filter(t => !t.completed);
    const completed = todos.filter(t => t.completed);
    
    let markdown = '# TODO Items Export\n\n';
    
    if (active.length > 0) {
      markdown += '## Active TODOs\n\n';
      active.forEach(todo => {
        const priorityEmoji = todo.priority === 'high' ? '🔴' : todo.priority === 'medium' ? '🟡' : '🟢';
        markdown += `### ${priorityEmoji} ${todo.title}\n\n`;
        markdown += `**ID:** ${todo.id}\n`;
        markdown += `**Priority:** ${todo.priority}\n`;
        markdown += `**Created:** ${new Date(todo.createdAt).toLocaleDateString()}\n`;
        if (todo.files && todo.files.length > 0) {
          markdown += `**Files:** ${todo.files.join(', ')}\n`;
        }
        if (todo.dueDate) {
          markdown += `**Due Date:** ${new Date(todo.dueDate).toLocaleDateString()}\n`;
        }
        markdown += `**Review:** ${todo.reviewId}\n\n`;
        if (todo.description && todo.description !== todo.title) {
          markdown += `${todo.description}\n\n`;
        }
        if (todo.example) {
          markdown += `**Example:**\n\`\`\`\n${todo.example}\n\`\`\`\n\n`;
        }
        markdown += '---\n\n';
      });
    }

    if (completed.length > 0) {
      markdown += '## Completed TODOs\n\n';
      completed.forEach(todo => {
        markdown += `- ✅ ~~${todo.title}~~ (${new Date(todo.createdAt).toLocaleDateString()})\n`;
      });
    }

    return markdown;
  }
}

File: core/todo_manager.py
"""TODO management system for tracking review suggestions."""

import json
from datetime import datetime, timedelta
from pathlib import Path
from typing import List, Optional, Dict, Any
from uuid import uuid4

from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from rich.progress import Progress, BarColumn, TaskProgressColumn
from rich import box

from models.types import TodoItem, Priority, Suggestion


class TodoManager:
    """Manages TODO items from code reviews."""
    
    def __init__(self, output_dir: Optional[Path] = None):
        """Initialize TODO manager.
        
        Args:
            output_dir: Directory for TODO storage
        """
        self.output_dir = output_dir or Path("reviews")
        self.todo_dir = self.output_dir / "todo"
        self.todo_dir.mkdir(parents=True, exist_ok=True)
        
        self.console = Console()
    
    async def get_all_todos(self) -> List[TodoItem]:
        """Get all TODO items.
        
        Returns:
            List of all TodoItem objects
        """
        todos = []
        
        for todo_file in self.todo_dir.glob("todo_*.json"):
            try:
                data = json.loads(todo_file.read_text(encoding='utf-8'))
                
                # Handle both single todo and list of todos
                if isinstance(data, list):
                    for item in data:
                        todos.append(self._parse_todo_item(item))
                else:
                    todos.append(self._parse_todo_item(data))
                    
            except (json.JSONDecodeError, KeyError) as e:
                self.console.print(f"[yellow]Warning: Failed to load {todo_file}: {e}[/yellow]")
                continue
        
        # Sort by priority and creation date
        priority_order = {Priority.HIGH: 0, Priority.MEDIUM: 1, Priority.LOW: 2}
        todos.sort(
            key=lambda t: (
                t.completed,  # Incomplete first
                priority_order.get(t.priority, 999),
                t.created_at
            )
        )
        
        return todos
    
    async def get_active_todos(self) -> List[TodoItem]:
        """Get only active (not completed) TODO items.
        
        Returns:
            List of active TodoItem objects
        """
        all_todos = await self.get_all_todos()
        return [todo for todo in all_todos if not todo.completed]
    
    async def get_completed_todos(self) -> List[TodoItem]:
        """Get only completed TODO items.
        
        Returns:
            List of completed TodoItem objects
        """
        all_todos = await self.get_all_todos()
        return [todo for todo in all_todos if todo.completed]
    
    async def get_todo_by_id(self, todo_id: str) -> Optional[TodoItem]:
        """Get a specific TODO by ID.
        
        Args:
            todo_id: TODO item ID
            
        Returns:
            TodoItem if found, None otherwise
        """
        all_todos = await self.get_all_todos()
        for todo in all_todos:
            if todo.id == todo_id:
                return todo
        return None
    
    async def mark_completed(self, todo_id: str) -> bool:
        """Mark a TODO item as completed.
        
        Args:
            todo_id: TODO item ID
            
        Returns:
            True if successful, False otherwise
        """
        todo = await self.get_todo_by_id(todo_id)
        if not todo:
            self.console.print(f"[red]TODO item '{todo_id}' not found[/red]")
            return False
        
        if todo.completed:
            self.console.print(f"[yellow]TODO item '{todo_id}' is already completed[/yellow]")
            return True
        
        todo.completed = True
        await self._update_todo(todo)
        
        self.console.print(f"[green]✓ Marked TODO '{todo.title}' as completed[/green]")
        return True
    
    async def mark_active(self, todo_id: str) -> bool:
        """Mark a TODO item as active (not completed).
        
        Args:
            todo_id: TODO item ID
            
        Returns:
            True if successful, False otherwise
        """
        todo = await self.get_todo_by_id(todo_id)
        if not todo:
            self.console.print(f"[red]TODO item '{todo_id}' not found[/red]")
            return False
        
        if not todo.completed:
            self.console.print(f"[yellow]TODO item '{todo_id}' is already active[/yellow]")
            return True
        
        todo.completed = False
        await self._update_todo(todo)
        
        self.console.print(f"[blue]↺ Marked TODO '{todo.title}' as active[/blue]")
        return True
    
    async def delete_todo(self, todo_id: str) -> bool:
        """Delete a TODO item.
        
        Args:
            todo_id: TODO item ID
            
        Returns:
            True if successful, False otherwise
        """
        todo = await self.get_todo_by_id(todo_id)
        if not todo:
            self.console.print(f"[red]TODO item '{todo_id}' not found[/red]")
            return False
        
        # Find and update the file containing this TODO
        for todo_file in self.todo_dir.glob("todo_*.json"):
            try:
                data = json.loads(todo_file.read_text(encoding='utf-8'))
                
                if isinstance(data, list):
                    filtered_data = [item for item in data if item.get('id') != todo_id]
                    if len(filtered_data) < len(data):
                        if filtered_data:
                            todo_file.write_text(json.dumps(filtered_data, indent=2), encoding='utf-8')
                        else:
                            todo_file.unlink()  # Delete empty file
                        
                        self.console.print(f"[red]🗑 Deleted TODO '{todo.title}'[/red]")
                        return True
                        
            except (json.JSONDecodeError, KeyError):
                continue
        
        return False
    
    async def update_todo(
        self,
        todo_id: str,
        title: Optional[str] = None,
        description: Optional[str] = None,
        priority: Optional[Priority] = None,
        assignee: Optional[str] = None,
        due_date: Optional[datetime] = None
    ) -> bool:
        """Update a TODO item.
        
        Args:
            todo_id: TODO item ID
            title: New title
            description: New description
            priority: New priority
            assignee: New assignee
            due_date: New due date
            
        Returns:
            True if successful, False otherwise
        """
        todo = await self.get_todo_by_id(todo_id)
        if not todo:
            self.console.print(f"[red]TODO item '{todo_id}' not found[/red]")
            return False
        
        if title:
            todo.title = title
        if description:
            todo.description = description
        if priority:
            todo.priority = priority
        if assignee is not None:
            todo.assignee = assignee
        if due_date is not None:
            todo.due_date = due_date
        
        await self._update_todo(todo)
        
        self.console.print(f"[green]Updated TODO '{todo.title}'[/green]")
        return True
    
    async def create_todo(
        self,
        title: str,
        description: str,
        priority: Priority = Priority.MEDIUM,
        files: Optional[List[str]] = None,
        assignee: Optional[str] = None,
        due_date: Optional[datetime] = None
    ) -> TodoItem:
        """Create a new TODO item.
        
        Args:
            title: TODO title
            description: TODO description
            priority: Priority level
            files: Related files
            assignee: Assigned to
            due_date: Due date
            
        Returns:
            Created TodoItem
        """
        todo = TodoItem(
            id=f"todo-{uuid4().hex[:8]}",
            title=title,
            description=description,
            priority=priority,
            files=files or [],
            completed=False,
            created_at=datetime.now(),
            review_id=f"manual-{datetime.now().strftime('%Y%m%d')}",
            assignee=assignee,
            due_date=due_date
        )
        
        # Save to file
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        todo_file = self.todo_dir / f"todo_manual_{timestamp}.json"
        
        todo_data = self._serialize_todo(todo)
        todo_file.write_text(json.dumps([todo_data], indent=2), encoding='utf-8')
        
        self.console.print(f"[green]Created TODO: {todo.title}[/green]")
        return todo
    
    async def display_todos(
        self,
        show_completed: bool = False,
        filter_priority: Optional[Priority] = None,
        filter_assignee: Optional[str] = None,
        search: Optional[str] = None
    ) -> None:
        """Display TODO items in a formatted table.
        
        Args:
            show_completed: Include completed TODOs
            filter_priority: Filter by priority
            filter_assignee: Filter by assignee
            search: Search in title and description
        """
        todos = await self.get_all_todos()
        
        # Apply filters
        if not show_completed:
            todos = [t for t in todos if not t.completed]
        
        if filter_priority:
            todos = [t for t in todos if t.priority == filter_priority]
        
        if filter_assignee:
            todos = [t for t in todos if t.assignee == filter_assignee]
        
        if search:
            search_lower = search.lower()
            todos = [
                t for t in todos
                if search_lower in t.title.lower() or search_lower in t.description.lower()
            ]
        
        if not todos:
            self.console.print("[yellow]No TODO items found[/yellow]")
            return
        
        # Create table
        table = Table(
            title="📋 TODO Items",
            box=box.ROUNDED,
            show_lines=True,
            title_style="bold cyan"
        )
        
        table.add_column("ID", style="dim", width=12)
        table.add_column("Status", width=8)
        table.add_column("Priority", width=8)
        table.add_column("Title", width=30)
        table.add_column("Description", width=40)
        table.add_column("Created", width=10)
        table.add_column("Assignee", width=10)
        
        for todo in todos:
            status = "✅" if todo.completed else "⭕"
            
            priority_style = {
                Priority.HIGH: "red",
                Priority.MEDIUM: "yellow",
                Priority.LOW: "green"
            }.get(todo.priority, "white")
            
            priority_icon = {
                Priority.HIGH: "🔴",
                Priority.MEDIUM: "🟡",
                Priority.LOW: "🟢"
            }.get(todo.priority, "⚪")
            
            # Truncate description
            desc = todo.description[:37] + "..." if len(todo.description) > 40 else todo.description
            
            table.add_row(
                todo.id,
                status,
                f"{priority_icon} {todo.priority.value}",
                todo.title,
                desc,
                todo.created_at.strftime("%Y-%m-%d"),
                todo.assignee or "-"
            )
        
        self.console.print(table)
        
        # Show summary
        active_count = sum(1 for t in todos if not t.completed)
        completed_count = sum(1 for t in todos if t.completed)
        
        self.console.print(
            f"\n[cyan]Total:[/cyan] {len(todos)} | "
            f"[yellow]Active:[/yellow] {active_count} | "
            f"[green]Completed:[/green] {completed_count}"
        )
    
    async def display_progress(self) -> None:
        """Display TODO completion progress."""
        todos = await self.get_all_todos()
        
        if not todos:
            self.console.print("[yellow]No TODO items to show progress[/yellow]")
            return
        
        # Calculate statistics
        total = len(todos)
        completed = sum(1 for t in todos if t.completed)
        active = total - completed
        
        # Priority breakdown
        high_count = sum(1 for t in todos if t.priority == Priority.HIGH and not t.completed)
        medium_count = sum(1 for t in todos if t.priority == Priority.MEDIUM and not t.completed)
        low_count = sum(1 for t in todos if t.priority == Priority.LOW and not t.completed)
        
        # Create progress bar
        progress = Progress(
            "[progress.description]{task.description}",
            BarColumn(),
            TaskProgressColumn(),
            console=self.console
        )
        
        with progress:
            task = progress.add_task(
                "[cyan]Overall Progress",
                total=total,
                completed=completed
            )
            progress.update(task, completed=completed)
        
        # Display statistics panel
        stats_content = f"""
[bold]📊 TODO Statistics[/bold]

Total TODOs: {total}
Completed: {completed} ({completed/total*100:.1f}%)
Active: {active}

[bold]Priority Breakdown (Active):[/bold]
🔴 High: {high_count}
🟡 Medium: {medium_count}
🟢 Low: {low_count}
"""
        
        panel = Panel(
            stats_content.strip(),
            title="Progress Report",
            border_style="cyan",
            padding=(1, 2)
        )
        
        self.console.print(panel)
    
    async def export_todos(
        self,
        format: str = "json",
        output_path: Optional[Path] = None
    ) -> Path:
        """Export TODO items to file.
        
        Args:
            format: Export format (json, csv, markdown)
            output_path: Output file path
            
        Returns:
            Path to exported file
        """
        todos = await self.get_all_todos()
        
        if not todos:
            raise ValueError("No TODO items to export")
        
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        
        if not output_path:
            filename = f"todos_export_{timestamp}.{format}"
            output_path = Path(filename)
        
        if format == "json":
            data = [self._serialize_todo(todo) for todo in todos]
            output_path.write_text(json.dumps(data, indent=2), encoding='utf-8')
            
        elif format == "csv":
            import csv
            
            with open(output_path, 'w', newline='', encoding='utf-8') as f:
                writer = csv.DictWriter(
                    f,
                    fieldnames=['id', 'title', 'description', 'priority', 'status', 
                               'created_at', 'assignee', 'due_date', 'files']
                )
                writer.writeheader()
                
                for todo in todos:
                    writer.writerow({
                        'id': todo.id,
                        'title': todo.title,
                        'description': todo.description,
                        'priority': todo.priority.value,
                        'status': 'completed' if todo.completed else 'active',
                        'created_at': todo.created_at.isoformat(),
                        'assignee': todo.assignee or '',
                        'due_date': todo.due_date.isoformat() if todo.due_date else '',
                        'files': ', '.join(todo.files)
                    })
                    
        elif format == "markdown":
            lines = ["# TODO Items Export", ""]
            lines.append(f"*Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}*", "")
            
            # Active TODOs
            active_todos = [t for t in todos if not t.completed]
            if active_todos:
                lines.append("## 📋 Active TODOs")
                lines.append("")
                
                for priority in [Priority.HIGH, Priority.MEDIUM, Priority.LOW]:
                    priority_todos = [t for t in active_todos if t.priority == priority]
                    if priority_todos:
                        icon = {"high": "🔴", "medium": "🟡", "low": "🟢"}[priority.value]
                        lines.append(f"### {icon} {priority.value.title()} Priority")
                        lines.append("")
                        
                        for todo in priority_todos:
                            lines.append(f"- **[{todo.id}]** {todo.title}")
                            lines.append(f"  - {todo.description}")
                            if todo.assignee:
                                lines.append(f"  - Assignee: {todo.assignee}")
                            if todo.due_date:
                                lines.append(f"  - Due: {todo.due_date.strftime('%Y-%m-%d')}")
                            lines.append("")
            
            # Completed TODOs
            completed_todos = [t for t in todos if t.completed]
            if completed_todos:
                lines.append("## ✅ Completed TODOs")
                lines.append("")
                
                for todo in completed_todos:
                    lines.append(f"- ~~[{todo.id}] {todo.title}~~")
                lines.append("")
            
            # Statistics
            lines.append("## 📊 Statistics")
            lines.append("")
            lines.append(f"- Total: {len(todos)}")
            lines.append(f"- Active: {len(active_todos)}")
            lines.append(f"- Completed: {len(completed_todos)}")
            lines.append(f"- Completion Rate: {len(completed_todos)/len(todos)*100:.1f}%")
            
            output_path.write_text('\n'.join(lines), encoding='utf-8')
            
        else:
            raise ValueError(f"Unsupported export format: {format}")
        
        self.console.print(f"[green]TODOs exported to: {output_path}[/green]")
        return output_path
    
    async def get_overdue_todos(self) -> List[TodoItem]:
        """Get overdue TODO items.
        
        Returns:
            List of overdue TodoItem objects
        """
        todos = await self.get_active_todos()
        now = datetime.now()
        
        return [
            todo for todo in todos
            if todo.due_date and todo.due_date < now
        ]
    
    async def get_upcoming_todos(self, days: int = 7) -> List[TodoItem]:
        """Get upcoming TODO items within specified days.
        
        Args:
            days: Number of days to look ahead
            
        Returns:
            List of upcoming TodoItem objects
        """
        todos = await self.get_active_todos()
        now = datetime.now()
        future = now + timedelta(days=days)
        
        return [
            todo for todo in todos
            if todo.due_date and now <= todo.due_date <= future
        ]
    
    def _parse_todo_item(self, data: Dict[str, Any]) -> TodoItem:
        """Parse TODO item from dictionary.
        
        Args:
            data: TODO data dictionary
            
        Returns:
            TodoItem object
        """
        return TodoItem(
            id=data['id'],
            title=data['title'],
            description=data['description'],
            files=data.get('files', []),
            priority=Priority(data.get('priority', 'medium')),
            completed=data.get('completed', False),
            example=data.get('example'),
            created_at=datetime.fromisoformat(data.get('created_at', datetime.now().isoformat())),
            review_id=data.get('review_id', 'unknown'),
            assignee=data.get('assignee'),
            due_date=datetime.fromisoformat(data['due_date']) if data.get('due_date') else None
        )
    
    def _serialize_todo(self, todo: TodoItem) -> Dict[str, Any]:
        """Serialize TODO item to dictionary.
        
        Args:
            todo: TodoItem object
            
        Returns:
            Serialized dictionary
        """
        return {
            'id': todo.id,
            'title': todo.title,
            'description': todo.description,
            'files': todo.files,
            'priority': todo.priority.value,
            'completed': todo.completed,
            'example': todo.example,
            'created_at': todo.created_at.isoformat(),
            'review_id': todo.review_id,
            'assignee': todo.assignee,
            'due_date': todo.due_date.isoformat() if todo.due_date else None
        }
    
    async def _update_todo(self, todo: TodoItem) -> None:
        """Update TODO item in storage.
        
        Args:
            todo: TodoItem to update
        """
        # Find the file containing this TODO
        for todo_file in self.todo_dir.glob("todo_*.json"):
            try:
                data = json.loads(todo_file.read_text(encoding='utf-8'))
                
                if isinstance(data, list):
                    for i, item in enumerate(data):
                        if item.get('id') == todo.id:
                            data[i] = self._serialize_todo(todo)
                            todo_file.write_text(json.dumps(data, indent=2), encoding='utf-8')
                            return
                            
            except (json.JSONDecodeError, KeyError):
                continue
        
        # If not found, create new file
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        todo_file = self.todo_dir / f"todo_updated_{timestamp}.json"
        todo_file.write_text(json.dumps([self._serialize_todo(todo)], indent=2), encoding='utf-8')